R/to_xml.R

```diff
@@ -10,17 +10,24 @@
 #'   preserved as best as possible. If this is `FASLE`, the anchors disappear
 #'   and the links will appear as normal links. See [resolve_anchor_links()] for
 #'   details.
+#' @param unescaped if `TRUE` (default) AND `sourcepos = TRUE`, square braces
+#'   that were unescaped in the original document will be presered as best as
+#'   possible. If this is `FALSE`, these braces will be escaped in the output
+#'   document. See [protect_unescaped()] for details.
 #'
 #' @return A list containing the YAML of the file (yaml)
-#' and its body (body) as XML.
+#'   and its body (body) as XML.
 #' 
 #' @details This function will take a (R)markdown file, split the yaml header
 #'   from the body, and read in the body through [commonmark::markdown_xml()].
 #'   Any RMarkdown code fences will be parsed to expose the chunk options in
 #'   XML and tickboxes (aka checkboxes) in GitHub-flavored markdown will be
 #'   preserved (both modifications from the commonmark standard). 
 #'
-#'   Math elements 
+#' @note
+#'   Math elements are not protected by default. You can use [protect_math()] to
```

Comment:
should there be an example / test of "apply all possible protections"

Reply:
Oh yes absolutely!

We should do a comparison of _everything off_ and then gradually build up to _everything on_. Of course the `yarn` implementation would be much simpler, because it can hide the complexity and allow folks to chain the protections. 

```r
source <- readLines(path)[-(1:3)] # source without yaml

# unescaped protection not possible
A <- to_xml(path, anchor_links = FALSE, unescaped = FALSE)
B <- to_xml(path, anchor_links = TRUE, unescaped = FALSE)

# equivalent to B
A$body <- anchor_links(A$body, source)

# unescaped protection possible
C <- to_xml(path, anchor_links = FALSE, unescaped = FALSE, sourcepos = TRUE)
D <- to_xml(path, anchor_links = TRUE, unescaped = TRUE, sourcepos = TRUE)

# equivalent to D
C$body <- anchor_links(C$body, source)
C$body <- protect_unescaped(C$body, source)
```

I'm realising that all of these are kind of 'off-label' enhancements we offer and I need to write a vignette that spells them out (including what happens with code chunks). The reason `anchor_links` and the `unescaped` are default is that they need the source document in order to process the elements that were hidden by the parser. The reason `protect_math` is not a default is because it is still not fully reliable as it will bork things that look like math (no matter how much I tried to avoid that).